File: lintcode/python/1780_the_candidate_with_the_most_votes.py
#!/usr/bin/python -t

# hash table

class Solution:
    """
    @param votes: The array of names of candidates in the election.
    @return: The name of the candidate who has the most votes.
    """
    def candidateWithTheMostVotes(self, votes):
        # Write your code here
        d = {}
        ret = ""
        max_cnt = 0
        
        for v in votes:
            d[v] = d.get(v, 0) + 1
            if d[v] >= max_cnt:
                max_cnt = d[v]
               
        for k, v in d.items():
            if v == max_cnt and (k < ret or len(ret) == 0):
                ret = k
                
        return ret
        

File: lintcode/golang/1780_the_candidate_with_the_most_votes.go
/**
 * @param votes: The array of names of candidates in the election.
 * @return: The name of the candidate who has the most votes.
 */

package main

import "fmt"

func main(){
   votes := []string{"John", "Johnny", "Jackie", "Johnny", "John", "Jackie", "Jamie", "Jamie", "John","Johnny", "Jamie", "Johnny", "John"}

    fmt.Println("best is ", candidateWithTheMostVotes(votes))
}

func candidateWithTheMostVotes (votes []string) string {
    // Write your code here
    b := make(map[string]int)

    for _, v := range votes {
        fmt.Println(v)
        if val, ok := b[v]; ok {
            b[v] = val+1
        } else {
            b[v] = 1
        }
    }

    fmt.Println(b)
    maxCnt := 0
    var ret string
    for k, v := range b {
        if v > maxCnt || (v == maxCnt && k < ret) || ret == "" {
            maxCnt = v
            ret = k
        }
    }

    return ret
}

